lib/multi.c

```diff
@@ -1116,6 +1116,68 @@ CURLMcode curl_multi_fdset(struct Curl_multi *multi,
   return CURLM_OK;
 }
 
+CURLMcode curl_multi_pollfds(struct Curl_multi *multi,
+                             struct curl_waitfd *ufds,
+                             unsigned int size,
+                             unsigned int *fd_count)
+{
+  struct Curl_easy *data;
+  unsigned int nfds = 0;
+  struct easy_pollset ps;
+  unsigned int i;
+  CURLMcode result = CURLM_OK;
+  struct curl_waitfd *ufd;
+  unsigned int j;
+
+  if(!GOOD_MULTI_HANDLE(multi))
+    return CURLM_BAD_HANDLE;
+
+  if(multi->in_callback)
+    return CURLM_RECURSIVE_API_CALL;
+
+  memset(&ps, 0, sizeof(ps));
+  for(data = multi->easyp; data; data = data->next) {
+    multi_getsock(data, &ps);
+
+    for(i = 0; i < ps.num; i++) {
+      if(nfds < size) {
+        curl_socket_t fd = ps.sockets[i];
+        int fd_idx = -1;
+
+        /* Simple linear search to skip addition of already added descriptor */
+        for(j = 0; j < nfds; j++) {
+          if(ufds[j].fd == fd) {
+            fd_idx = (int)j;
+            break;
+          }
+        }
+
+        if(fd_idx < 0) {
+          ufd = &ufds[nfds++];
+          ufd->fd = ps.sockets[i];
+          ufd->events = 0;
+        }
+        else
+          ufd = &ufds[fd_idx];
+
+        if(ps.actions[i] & CURL_POLL_IN)
+          ufd->events |= CURL_WAIT_POLLIN;
+        if(ps.actions[i] & CURL_POLL_OUT)
+          ufd->events |= CURL_WAIT_POLLOUT;
+      }
+      else if(size > 0) {
```

Comment:
Why this check? Surely size 0 is also too small?

Reply:
Sorry, I was off for a while.
This check allows just to get a number of descriptors when the passed 'size' is 0 and not return an error when the size is too small (which for size 0 will happen on the first descriptor).

In other words, it provides client an option to pass the `size 0` to get the number of "wait" descriptors, allocate necessary storage for them, and get them into that storage in the next call. This helps to avoid memory reallocations if they are critical.

Reply:
Sure, I will try to create a man page for this function using the previous man pages (i.e. curl_multi_fdset.3), but I may need some help as I am not familiar with the man page format.

Reply:
@bagder Could you please look at the man page and tell me what is wrong, so I can fix it?

I don't understand why I get the spelling error:
Misspelled words:
<htmlcontent> docs/libcurl/curl_multi_waitfds.md: html>body>p
--------------------------------------------------------------------------------
waitfds
--------------------------------------------------------------------------------

I looked at the other pages, and I don't see why I am getting this error.

Reply:
Try this:
~~~diff
diff --git a/.github/scripts/cleanspell.pl b/.github/scripts/cleanspell.pl
index 329a97240..d8475a819 100755
--- a/.github/scripts/cleanspell.pl
+++ b/.github/scripts/cleanspell.pl
@@ -63,11 +63,11 @@ while(<F>) {
         $_ =~ s/CURLUPART_[A-Z0-9_]*//g;
         $_ =~ s/CURLUSESSL_[A-Z0-9_]*//g;
         $_ =~ s/curl_global_(init_mem|sslset|cleanup)//g;
         $_ =~ s/curl_(strequal|strnequal|formadd|waitfd|formget|getdate|formfree)//g;
         $_ =~ s/curl_easy_(nextheader|duphandle)//g;
-        $_ =~ s/curl_multi_fdset//g;
+        $_ =~ s/curl_multi_[a-z]*//g;
         $_ =~ s/curl_mime_(subparts|addpart|filedata|data_cb)//g;
         $_ =~ s/curl_ws_(send|recv|meta)//g;
         $_ =~ s/curl_url_(dup)//g;
         $_ =~ s/curl_pushheader_by(name|num)//g;
         $_ =~ s/libcurl-(env|ws)//g;

Reply:
@bagder Thanks! I think I resolved all the issue, except one, which is related to documentation.
The failing check log says:
```
...
 curl_multi_strerror.3
 curl_multi_timeout.3
 curl_multi_wait.3
-curl_multi_waitfds.3
 curl_multi_wakeup.3
 curl_pushheader_byname.3
 curl_pushheader_bynum.3
ERROR: installed libcurl docs mismatch
Error: Process completed with exit code 2.

```
So, I think I need to modify one more script and add `curl_multi_waitfds` somewhere, right?

Reply:
@bagder As I can see these changes didn't get into 8.6.0,  I am wondering if it is possible to  get them into the next release?